### ISSUE TITLE
Better explain how pure is different from side effect free

### ISSUE THREAD

**Issue by calvinmetcalf** [js-compiler-hints-spec/compiler-hint-notations-spec-draft#3]
It took me a couple readings to realize pure no use of closures. Making it more explicit what the two things do would by helpful. My understanding:

- side effect free: does not write to any variable outside the function and does not modify it's arguments if they are objects
- pure: same as above but also does not read from any external variables besides it's arguments

**Comment by antfu** [comment]
I agree. But I feel that the problem is that `__PURE__` is used on expressions, but not on function declarations, which does not apply to the statement you mentioned?

From my understanding:

- `__PURE__` marks expressions as "side-effect-free"
- `__NO_SIDE_EFFECTS__` marks functions as "side-effect-free"

So there is actually no "pure function" concept involved.

**Comment by robpalme** [comment]
The semantic relationship I have always understood is:

- **side effect:**  Calling this function may mutate state outside the function, e.g. mutation of non-primitive arguments
- **side cause:**  This function's return value may depend on state outside the function, e.g. a closure-captured variable
- **pure:**  This function has no _side effects_ and no _side causes_.

I agree this relation should be clarified somewhere.
